OpenRA.Mods.Common/Activities/Air/Fly.cs

```diff
@@ -149,16 +149,22 @@ public override bool Tick(Actor self)
 			if (!targetIsHiddenActor && target.Type == TargetType.Actor)
 				lastVisibleTarget = Target.FromTargetPositions(target);
 
-			useLastVisibleTarget = targetIsHiddenActor || !target.IsValidFor(self);
```

Comment:
It has been too long since I last thought about the target tracking code in detail, and it took a concerted effort to get back into the mindset and remember the original reasoning behind this.

The idea behind this line was to separate the various cases into two bins: the target is an actor that is currently visible (&rarr; `useLastVisibleTarget == false`) vs "we have no knowledge of the actor state this tick" (&rarr; `useLastVisibleTarget == true`).

There are a bunch of different cases in this second bin*, but we define the same behavior for all of them: if we don't know the current location, move to the last known location. We will either recover vision in a future tick, or we'll reach the requested range of the last known target and the activity will end.

Note that this also covers the case where we saw the target die: we defined / decided as an intentional gameplay mechanic that units should continue moving within range of where the target died - in part because it makes the code so much simpler, but also because we felt it better for gameplay than units going idle at arbitrary times, potentially while in the middle of nowhere while moving long range.

This logic *should* work fine provided we can guarantee that `lastVisibleTarget` has a sensible value. If the activity ticks at least once with the target visible then we are fine. Otherwise, we rely on the information passed into the activity constructor. If we have neither there is no way to recover, and the `useLastVisibleTarget && !lastVisibleTarget.IsValidFor(self)` bails us out.

*Off the top of my head, there may be more:
* Target is (alive or dead), but under the fog
* Target is (alive or dead), but represented by a frozen actor
* Target is dead, and we saw it die
* Target is alive, but conditions / transforming mean we can no target it (i had previously forgotten about this one)

Reply:
With this context in place, can you outline the goal behaviour with the changes here?

If the main motivation is to make units go idle when the target dies, then that is actually a significant change that we need to consider in more detail and should be applying uniformly to all units rather than just aircraft.

Reply:
The context was pretty clear already, but thanks anyway.

We need to bail out of the activity as soon as the target is dead in order to address #16922. The change here makes the plane exit the activity when the target is known to have died (e.g. when the plane itself killed it), but move to the last known location if it's fate is unknown (and then then exit anyway if it cannot be found).

Yes, this also makes the plane exit the activity early when something else visibly killed it. This was not the motivation for the change here, but it seems a lot more sensible than the alternative anyway. There is no reason to go to that specific location and then idle anyway when the specific target of a focus fire order is already dead. If the player explicitly wanted the plane to go there regardless, they should have used an attack-move instead.

Yes, this should be changed for all the other activities that use this targetting logic as well for consistency, but that is not needed for the actual bugfix here.

Reply:
> There is no reason to go to that specific location and then idle anyway when the specific target of a focus fire order is already dead.

The main reason is that this is the way all the other units have worked for the last couple of releases. Can we please introduce a `cancelOnTargetVisiblyRemoved` (ideally with a better variable name) flag to the `Fly` ctor so we can enable this behaviour when queued as a child of `FlyAttack`, but minimizes behaviour changes for anything else that queues it?

> This was not the motivation for the change here, but it seems a lot more sensible than the alternative anyway.

This remains to be debated, ideally informed by examples from how other games handle it. I'm not convinced that units going idle at an arbitrary time/place (which may not be safe: e.g. infantry on Tiberium) and requiring the player to recover them is a better solution than completing the move to a location that probably has other enemy targets nearby.

Reply:
I'd rather not get dragged into any 'debates' again.